### ISSUE TITLE
pachd Looking to Scale Up Workers Every 2 Seconds

### ISSUE THREAD

**Issue by dgeorg42** [pachyderm/pachyderm#6582]
After creating a pipeline called "benchmarking", when I look at the `pachd` logs, I see the following messages logged every 2 seconds:
```
2021-07-21T21:45:47Z INFO PPS master: processing event for "benchmarking"
2021-07-21T21:45:47Z INFO pfs.API.GetFileTAR {"request":{"file":{"commit":{"branch":{"repo":{"name":"benchmarking","type":"spec"},"name":"master"},"id":"7a732661c11e426c9dbe81ac7b0154d8"},"path":"spec"}}} 
2021-07-21T21:45:47Z INFO pfs.API.GetFileTAR {"duration":0.013665573,"request":{"file":{"commit":{"branch":{"repo":{"name":"benchmarking","type":"spec"},"name":"master"},"id":"7a732661c11e426c9dbe81ac7b0154d8"},"path":"spec"}}} 
2021-07-21T21:45:47Z INFO PPS master: scaling up workers for "benchmarking"
```
If I'm interpreting these messages correctly, there's an event every 2 seconds which causes `pachd` to query the `spec` repo for the pipeline and then determine if it needs to scale up workers.  Which would make sense, except that there are no jobs running, there aren't even any files in the input repo, the pipeline doesn't have autoscaling enabled, and there's already a single worker available, so it's not clear to me why this event would need to be triggered at all, let alone be triggered every 2 seconds.

**Comment by jrockway** [comment]
I noticed this too and might have a bug open.  I think this is just a misleading log message.  The message is printed and the function can then decide to not do any work.

```go
func (op *pipelineOp) scaleUpPipeline() (retErr error) {
        log.Infof("PPS master: scaling up workers for %q", op.pipelineInfo.Pipeline.Name)
        ...
        // update pipeline RC
        return op.updateRC(func(rc *v1.ReplicationController) {
                if rc.Spec.Replicas != nil && *op.rc.Spec.Replicas > 0 {
                        return // prior attempt succeeded
                }
       ...
```

You can look at the http_client_requests_total Prometheus counter to see if it's actually trying to update kubernetes every few seconds.  If it is, that is problematic.

**Comment by jrockway** [comment]
What triggers this operation is reading an event from m.eventCh, and the code writes this event every 2 seconds, so it seems like it's intentional.  Commented and everything ;)

poller.go
```go
                // generate a pipeline event for 'pipeline'
                log.Debugf("PPS master: polling pipeline %q", pipeline)
                select {
                case m.eventCh <- &pipelineEvent{eventType: writeEv, pipeline: pipeline}:
                        break
                case <-ctx.Done():
                        break
                }

                // 5. move to next pipeline (after 2s sleep)
                return nil
```

master.go
```go
                select {
                case e := <-m.eventCh:
                        switch e.eventType {
                        case writeEv:
                                if err := m.attemptStep(m.masterCtx, e); err != nil {
                                        log.Errorf("PPS master: %v", err)
                                }
               ...
```

attemptStep calls step which logs "processing event".  So all in all, this seems intentional, but verbose.

**Comment by dgeorg42** [comment]
On the one hand, I agree that it's just a misleading error message. Might be more accurate if it logged something like `PPS master: potentially scaling up workers for "benchmarking" if it turns out that we actually need to`. :)

On the other hand, though, the bigger question is: Why are we checking every 2 seconds to see if we need to scale up the pipeline's workers when there is literally nothing going on? No jobs running, nothing in the repo, autoscaling not even enabled, etc.  My concern is that these constant unnecessary checks are going to be problematic when we try to scale up to hundreds of pipelines.

**Comment by jdoliner** [comment]
@msteffen would you mind taking a look at this, or maybe just clarifying why we check at the rate we do? I believe this is part of the pipeline controller code. I don't think this is part of the autoscaling code but if I'm wrong about that let me know and I can take a look since I'm more familiar with that.

**Comment by jrockway** [comment]
I don't think this is related to autoscaling.  I'd change the Info log to debug ("considering xxx for scaling up"), and then log at the info level when the number in the RC actually changes.  (i guess it would be "scaling down" if the number decreases, but i feel like 1 -> 1 is the most common case, and 0 -> 1 is the second most common case.)

It would be most nice, probably, to export prometheus metrics for all the periodic work the PPS controller does.

**Comment by msteffen** [comment]
@dgeorg42 To your questions:

> On the other hand, though, the bigger question is: Why are we checking every 2 seconds to see if we need to scale up the pipeline's workers when there is literally nothing going on? No jobs running, nothing in the repo, autoscaling not even enabled, etc.

This code's purpose is to reconcile kubernetes resources that PPS otherwise may have dropped (e.g. if the kubernetes API server becomes unavailable for some period, and a pipeline is deleted during that period, but due to the unavailability the pipeline's RC is not deleted).

> My concern is that these constant unnecessary checks are going to be problematic when we try to scale up to hundreds of pipelines.

Per the characterization above, I disagree that these checks are unnecessary, though I'd be happy to discuss that point in detail. As far as performance implications though: the code doesn't check every pipeline every two seconds, it checks one pipeline every two seconds and then moves on to the next pipeline. If you have ten pipelines, it checks each pipeline once every twenty seconds. If you have hundreds of pipelines, the load on PFS/Kubernetes stays constant, but pipelines are checked less frequently. If a stray RC is left around, it'll take longer to clean up in that case, but hopefully that's acceptable as it's a lower proportion of overall PPS resources.

I think changing the log message could be helpful, and of course I'd be happy to go over this code in more detail.

**Comment by msteffen** [comment]
Per slack: the fix here seems to be making the `scaling up workers for "benchmarking"` log message a) more accurate (e.g. `checking resources for ...`) and b) log at the DEBUG level, so it doesn't overwhelm the logs

I'll try and make this fix today
